### PR TITLE
Display Notes in TODO View and Event/Appt. View

### PR DESCRIPTION
**Pull request by rosalogia**
This PR implements the feature requested in #289 and expands this functionality to events and appointments as well. When viewing TODO items or Events/Appts., if one exists, the associated note is displayed in the pop-up window and is truncated after reaching a specified character limit.

EDIT: I would like to be clear about the fact that I'm very inexperienced with C, and would appreciate any effort to review or further test the code from anyone willing to help :smile:

### CODE REVIEW

**Comment by lfos** [comment]
Thanks a lot Ambika! Your approach looks good overall but I do have a few comments regarding the implementation:
* Is there a reason we're limiting the note output to 3500 characters? What happens if we read the whole file?
* There seem to be whitespace/indentation issues; particularly in the asprintf() line. Please also make sure each commit is properly formatted on its own (as opposed to introducing inconsistent indentation and fixing it in a followup commit).
* Please try to avoid strcpy() and any buffer length computations. For example, instead of using snprintf() you can use asprintf() which automatically allocates a buffer that is large enough to hold the result. No careful computations needed.

**Comment by rosalogia** [comment]
Hi @lfos! Sorry for the late reply.

> Is there a reason we're limiting the note output to 3500 characters? What happens if we read the whole file?

Not particularly! But the pop-up window will automatically truncate text after a certain point (without a trailing `...`) so I felt that reading in the whole file might have been wasteful and introduce unnecessary complexity.

I found that it's quite difficult to predict at what point text will be automatically truncated, since it is not dependent at all on number of lines or length of lines. Instead, text automatically wraps, and one very very long line or many very very short lines will cause auto-truncation. I am not sure that there is a way to calculate how a hard maximum that actually corresponds to how text is rendered in the pop-up, so I went with a number that didn't truncate too early for large text files.

Of course, if we were to read in the whole file we could just let the pop-up window handle truncation, and although we would be missing the trailing `...`, they are arguably unnecessary. However, I am not sure whether or not this would be the most resource-efficient way of going about this.

> There seem to be whitespace/indentation issues; particularly in the asprintf() line.

I noticed this initially, and I'm not sure whether or not you noticed that in the second commit on the PR these issues were fixed (as far as I'm aware).

> Please also make sure each commit is properly formatted on its own (as opposed to introducing inconsistent indentation and fixing it in a followup commit).

So would you rather I rebase and squash these two commits into one so that a commit with improper formatting isn't introduced to the branch history? If so, that's something I can do!

> Please try to avoid strcpy() and any buffer length computations. For example, instead of using snprintf() you can use asprintf() which automatically allocates a buffer that is large enough to hold the result. No careful computations needed.

This is absolutely doable. Thanks!

**Comment by lfos** [comment]
> Of course, if we were to read in the whole file we could just let the pop-up window handle truncation, and although we would be missing the trailing `...`, they are arguably unnecessary. However, I am not sure whether or not this would be the most resource-efficient way of going about this.

Thanks for the explanation. There are probably more sophisticated ways of achieving the same thing in a resource efficient yet more generic/consistent way (e.g. using the product of the numbers of rows and the number of columns of the screen as a upper bound for the contents read from the file, then adding the ellipsis in the rendering code) but the hardcoded limit seems like a good start.

> > There seem to be whitespace/indentation issues; particularly in the asprintf() line.
> 
> I noticed this initially, and I'm not sure whether or not you noticed that in the second commit on the PR these issues were fixed (as far as I'm aware).

I did but it's always preferrable to keep commits atomic and clean (both for readers of the history and for certain Git commands operating on individual commits).

> > Please also make sure each commit is properly formatted on its own (as opposed to introducing inconsistent indentation and fixing it in a followup commit).
> 
> So would you rather I rebase and squash these two commits into one so that a commit with improper formatting isn't introduced to the branch history? If so, that's something I can do!

It should be fairly straightforward to uncommit your second commit (e.g. using `git reset HEAD^`), stage only the whitespace changes (e.g. using `git add -p`), then merge the whitespace changes into the first commit (e.g. using `git commit --amend`), then recreate the second commit only with the non-whitespace changes (e.g. using `git commit -a`).

> > Please try to avoid strcpy() and any buffer length computations. For example, instead of using snprintf() you can use asprintf() which automatically allocates a buffer that is large enough to hold the result. No careful computations needed.
> 
> This is absolutely doable. Thanks!

Thank you!

**Comment by rosalogia** [comment]
Apologies for the repeated force-pushes :grimacing: I'm not sure there's a better way to amend history that's been pushed already but I definitely could've minimised the quantity. Each time I thought I was done I noticed something else I had forgotten :sob:

That being said, I'm a little unsure as to how I might do away with the use of `strcpy` in [L165](https://github.com/lfos/calcurse/pull/322/commits/d2c64fe57f7765ebcd26fc6cd01890eedcfcb548#diff-e128ceb99ac8eb7d8ad1ec858b833fa1R165) of `note.c`. What I'm trying to achieve here is replace the final few elements of the buffer with `"...\0"` if the file exceeds maximum length. Could this be done with `asprintf` or a related function?

**Comment by lfos** [comment]
No worries. In that specific case, you could either use asprintf() with a length specifier to copy 3497 characters, then append the `...`, or use your approach; I'd suggest using memcpy() instead of strcpy() with what you're doing, just to make it more explicit that the command also adds the terminating NUL-character, though.

**Comment by JPTIZ** [comment]
Finally got my chance to review this :) I can see that there's been a lot of discussion here already. So, here: I tried to compile the code with your PR, but I'm stuck at this:

```
day.c: In function ‘day_popup_item’:
day.c:594:4: error: implicit declaration of function ‘note_read_contents’ [-Werror=implicit-function-declaration]
  594 |    note_read_contents(note, note_size, fp);
      |    ^~~~~~~~~~~~~~~~~~
```

(I have `-Werror=implicit-function-declaration` flag set globally for reasons of my work, which triggers this as an error, but it is originally a warning)

I would suggest to declare the function in `calcurse.h` (there's a list of function declarations for `ui-day.c` around line 1145).

**Comment by rosalogia** [comment]
I believe I have handled both of these concerns! Please let me know if there's anything I missed :smile:

File: src/ui-todo.c [hunk 1]
```diff
@@ -314,7 +314,28 @@ void ui_todo_popup_item(void)
...
-	item_in_popup(NULL, NULL, item->mesg, _("TODO:"));
+	if (item->note) {
+		/* Assign a sane default note size that will cleanly
+		 * truncate long notes */
+		size_t note_size = 3500;
+		char note[note_size];
+
+		char *notepath;
+		FILE *fp;
+		asprintf(&notepath, "%s%s", path_notes, item->note);
```
Comment (lfos) [inline src/ui-todo.c:325]: Memory leak, please run mem_free(notepath) when notepath is no longer needed.

File: src/ui-todo.c [hunk 2]
```diff
@@ -314,7 +314,28 @@ void ui_todo_popup_item(void)
...
+
+		fp = fopen(notepath, "r");
+		note_read_contents(note, note_size, fp);
+		fclose(fp);
+
+		char *msg;
+
+		asprintf(&msg, "%s\n\nNOTES:\n%s", item->mesg, note);
```
Comment (lfos) [inline src/ui-todo.c:333]: Memory leak, please run mem_free(msg).

File: src/ui-todo.c [hunk 3]
```diff
@@ -314,7 +314,28 @@ void ui_todo_popup_item(void)
...
+
+		item_in_popup(NULL, NULL, msg, _("TODO:"));
+	}
+	else
```
Comment (lfos) [inline src/ui-todo.c:337]: Style: Please put else on the same line as the closing braces; use braces for an else-block if the if-block has braces (even if the body consists of just a single line). This line should be "} else {".

File: src/day.c [hunk 4]
```diff
@@ -582,8 +582,28 @@ void day_write_stdout(time_t date, const char *fmt_apt, const char *fmt_rapt,
...
 	if (day->type == EVNT || day->type == RECUR_EVNT) {
-		item_in_popup(NULL, NULL, day_item_get_mesg(day),
-			      _("Event:"));
+		if (day_item_get_note(day)) {
+			size_t note_size = 3500;
+			char note[note_size];
+
+			char *notepath;
+			FILE *fp;
+			asprintf(&notepath, "%s%s", path_notes, day_item_get_note(day));
```
Comment (lfos) [inline src/day.c:591]: Memory leak. Please fix (see comments in previous patch for details).

File: src/day.c [hunk 5]
```diff
@@ -582,8 +582,28 @@ void day_write_stdout(time_t date, const char *fmt_apt, const char *fmt_rapt,
...
+
+			fp = fopen(notepath, "r");
+			note_read_contents(note, note_size, fp);
+			fclose(fp);
+
+			char *msg;
+
+			asprintf(&msg, "%s\n\nNOTES:\n%s", day_item_get_mesg(day), note);
```
Comment (lfos) [inline src/day.c:599]: Memory leak. Please fix.

File: src/day.c [hunk 6]
```diff
@@ -593,8 +613,28 @@ void day_popup_item(struct day_item *day)
...
-		item_in_popup(a_st, a_end, day_item_get_mesg(day),
-			      _("Appointment:"));
+
+		if (day_item_get_note(day)) {
+			size_t note_size = 3500;
+			char note[note_size];
+
+			char *notepath;
+			FILE *fp;
+			asprintf(&notepath, "%s%s", path_notes, day_item_get_note(day));
```
Comment (lfos) [inline src/day.c:623]: Memory leak. Please fix.

File: src/day.c [hunk 7]
```diff
@@ -593,8 +613,28 @@ void day_popup_item(struct day_item *day)
...
+
+			fp = fopen(notepath, "r");
+			note_read_contents(note, note_size, fp);
+			fclose(fp);
+
+			char *msg;
+
+			asprintf(&msg, "%s\n\nNOTES:\n%s", day_item_get_mesg(day), note);
```
Comment (lfos) [inline src/day.c:631]: Memory leak. Please fix.

File: src/day.c [hunk 8]
```diff
@@ -582,8 +582,28 @@ void day_write_stdout(time_t date, const char *fmt_apt, const char *fmt_rapt,
...
+
+			item_in_popup(NULL, NULL, msg,
+					_("Event:"));
```
Comment (lfos) [inline src/day.c:602]: Style: No line break, append to previous line.

File: src/day.c [hunk 9]
```diff
@@ -582,8 +582,28 @@ void day_write_stdout(time_t date, const char *fmt_apt, const char *fmt_rapt,
...
+		}
+		else
```
Comment (lfos) [inline src/day.c:604]: Style: Move else to line above and add braces (see previous patch for details).

File: src/day.c [hunk 10]
```diff
@@ -593,8 +613,28 @@ void day_popup_item(struct day_item *day)
...
+			note_read_contents(note, note_size, fp);
+			fclose(fp);
+
+			char *msg;
+
+			asprintf(&msg, "%s\n\nNOTES:\n%s", day_item_get_mesg(day), note);
+			item_in_popup(a_st, a_end, msg,
+					_("Appointment:"));
+		}
+		else
```
Comment (lfos) [inline src/day.c:635]: Style: Move else to line above and add braces (see previous patch for details).

File: src/day.c [hunk 11]
```diff
@@ -593,8 +613,28 @@ void day_popup_item(struct day_item *day)
...
+
+			fp = fopen(notepath, "r");
+			note_read_contents(note, note_size, fp);
+			fclose(fp);
+
+			char *msg;
+
+			asprintf(&msg, "%s\n\nNOTES:\n%s", day_item_get_mesg(day), note);
+			item_in_popup(a_st, a_end, msg,
+					_("Appointment:"));
```
Comment (lfos) [inline src/day.c:633]: Style: No line break, append to previous line.

**Review by lfos — COMMENTED** [review]
Thanks for the revision! I only have a few minor comments; see inline.

**Comment by rosalogia** [comment]
Thanks :smile: I hadn't consciously registered that once I started using `asprintf` I was allocating dynamically. Hopefully I've addressed all your concerns, but please do let me know if there's anything I missed!

File: src/ui-todo.c [hunk 12]
```diff
@@ -314,7 +314,30 @@ void ui_todo_popup_item(void)
...
+		 * truncate long notes */
+		size_t note_size = 3500;
+		char note[note_size];
+
+		char *notepath;
+		FILE *fp;
+		asprintf(&notepath, "%s%s", path_notes, item->note);
+
+		fp = fopen(notepath, "r");
+                mem_free(notepath);
```
Comment (lfos) [inline src/ui-todo.c:328]: Whitespace issue. Please fix.

File: src/day.c [hunk 13]
```diff
@@ -582,8 +582,30 @@ void day_write_stdout(time_t date, const char *fmt_apt, const char *fmt_rapt,
...
+			fclose(fp);
+
+			char *msg;
+
+			asprintf(&msg, "%s\n\nNOTES:\n%s", day_item_get_mesg(day), note);
+
+			item_in_popup(NULL, NULL, msg, _("Event:"));
+
+			mem_free(msg);
+		} else  {
```
Comment (lfos) [inline src/day.c:606]: Style nit: Duplicate space.

**Review by lfos — COMMENTED** [review]
I think we're getting really close. I found two more small whitespace issues. Looks good otherwise, thanks!

**Comment by rosalogia** [comment]
This should do it :) Thanks for all your help and of course do let me know if there's anything that could be fixed or improved upon!

**Comment by JPTIZ** [comment]
Yay, nice, seems to work perfectly here :)

image

**Comment by lfos** [comment]
Merged into pu, thanks!

(By the way, I noticed that your last name is different in the Author header of the two patches. Is that intended?)

**Comment by rosalogia** [comment]
It is not intended :sweat_smile: Some of the information in my local git configuration is/was outdated and I must have updated some of it while this PR was being worked on. It's not much of a bother to me, but if it's easy to fix maybe that's worth doing!

**Comment by lhca** [comment]
Sorry for this belated comment.

The text `NOTES:` should be translatable and not hard coded. Except for the word "TODO", which is special, capitalized forms are not used elsewhere. I would prefer "Note:" as default.

**Comment by lfos** [comment]
Agreed, good catch Lars. I amended the patches in `pu`.

**Comment by lfos** [comment]
Ambika, I also updated your author information in the first patch (assuming the information in the second patch was correct). If you would like to change it, please let me know.

**Pull request closed without merging**